src/ir/analysis/derive.rs

```diff
@@ -308,62 +340,90 @@ impl<'ctx> CannotDerive<'ctx> {
-                                    self.derive_trait
-                                );
+                            // fall through to be same as non-union handling
+                        }
+                        trait_can_derive => {
+                            if self.ctx.options().rust_features().untagged_union
+                            {
+                                match trait_can_derive {
+                                    CanDerive::Manually => {
+                                        trace!(
+                                            "    cannot derive {} for Rust union, but it can be implemented",
+                                            self.derive_trait
+                                            );
+                                    }
+                                    _ => {
+                                        trace!(
+                                            "    cannot derive {} for Rust unions",
+                                            self.derive_trait
+                                            );
+                                    }
+                                }
+                                return trait_can_derive;
                             }
-                        };
-                        return layout_can_derive;
+
+                            let layout_can_derive = ty.layout(self.ctx).map_or(
+                                CanDerive::Yes,
+                                |l| {
+                                    l.opaque().array_size_within_derive_limit(
+                                        self.ctx,
+                                    )
+                                },
+                            );
+                            match layout_can_derive {
+                                CanDerive::Yes => {
+                                    trace!(
+                                        "    union layout can trivially derive {}",
+                                        self.derive_trait
+                                    );
+                                }
+                                _ => {
+                                    trace!(
+                                        "    union layout cannot derive {}",
+                                        self.derive_trait
+                                    );
+                                }
+                            };
+                            return layout_can_derive;
+                        }
                     }
                 }
 
-                if !self.derive_trait.can_derive_compound_with_vtable() &&
-                    item.has_vtable(self.ctx)
-                {
-                    trace!(
-                        "    cannot derive {} for comp with vtable",
-                        self.derive_trait
-                    );
-                    return CanDerive::No;
+                match (
```

Comment:
I think this would be more readable like:

```rust
if item.has_vtable(self.ctx) {
    let can_derive = self.derive_trait.can_derive_compound_with_vtable();
    if can_derive != CanDerive::Yes {
        trace!(...);
        return can_derive;
    }
}
```

Reply:
Done.